### PR TITLE
feat: added functionality to test providers creds during creation

### PR DESCRIPTION
**Pull request by BiswaViraj**
-  **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
   - [x] Feature
-  **What is the current behavior?** (You can also link to an open issue here)
   - Currently, while adding a provider from the Integration store, there's no functionality to check/verify for potential bad/wrong credentials
- **What is the new behavior (if this is a feature change)?**
  - Added functionality to check for wrong creds while connecting a provider by sending a dummy Email using those creds.
  - Updated the `createIntegration` & `updateIntegration` API to accept `check: boolean` for handling checkIntegartion invockation.
  - Added a Switch toggle on the UI for `Email` channel only.
  
image
^ Might need to change the `Switch` text and design.

### CODE REVIEW

**Review by scopsy — COMMENTED** [review]
I think that when adding the first handler here, it's good to define the potential throw error messages to make sure that we can also show the user the provider raw error response, but for some common errors such as:

- From email is not validated
- Bad Credentials (i.e, API Key, etc...)
- Other

Also, we need to make sure we properly display the error returned from the provider, from our known error list, we in the future could also a guide from each provider on how to handle this error with a link to their docs. Wdyt?

File: apps/api/src/app/integrations/usecases/create-integration/create-integration.usecase.ts [hunk 1]
```diff
@@ -17,6 +20,15 @@ export class CreateIntegration {
     let response: IntegrationEntity;
 
     try {
+      const res = await this.checkIntegration.execute(
```
[the remark this answers is not in the document]
Comment (BiswaViraj) [reply apps/api/src/app/integrations/usecases/create-integration/create-integration.usecase.ts:23]: yes, also agree about the error msgs to create a generic one having all the common errors and provider-specific errors as you mentioned.

File: providers/sendgrid/src/lib/sendgrid.provider.ts [hunk 2]
```diff
@@ -38,11 +73,18 @@ export class SendgridEmailProvider implements IEmailProvider {
...
-    });
-
-    return {
-      id: response[0]?.headers['x-message-id'],
-      date: response[0]?.headers?.date,
     };
   }
 }
+
+const mapResponse = (statusCode: number) => {
```
Comment (BiswaViraj) [inline providers/sendgrid/src/lib/sendgrid.provider.ts:80]: Error codes mapping referenced from [Sendgrid API Responses](https://docs.sendgrid.com/api-reference/how-to-use-the-sendgrid-v3-api/responses)

File: apps/web/src/pages/integrations/components/ConnectIntegrationForm.tsx [hunk 3]
```diff
@@ -131,7 +143,25 @@ export function ConnectIntegrationForm({
...
+                    data-test-id="is_check_integration_id"
+                    {...field}
+                    onChange={handlerCheckIntegrationChange}
+                  />
+                )}
+              />
+              <StyledText>Send a test email to verify the credentials?</StyledText>
+            </CheckIntegrationWrapper>
+          )}
+        </Stack>
```
Comment (BiswaViraj) [inline apps/web/src/pages/integrations/components/ConnectIntegrationForm.tsx:163]: Currently, the option to test connection will be only enabled for create new integration and Email channel

**Comment by scopsy** [comment]
@oba2311 could you take a look at the design section of the switch box? I think it's a bit crowded there. What do you think? Maybe we can just enable it by default without the ability to skip the test? Even tho I think it might be in edge case a wanted behaviour to submit the form even if the test is failing. 

Wdyt?

File: apps/web/src/pages/integrations/components/ConnectIntegrationForm.tsx (hunk 3 above)
Comment (scopsy) [reply apps/web/src/pages/integrations/components/ConnectIntegrationForm.tsx:163]: Could we make it work also for "update" ?

**Comment by scopsy** [comment]
CleanShot 2022-09-18 at 15 51 24@2x


@BiswaViraj when clicking on the "Connect" button nothing happens, the error happened on the server is not displayed to the user. 

I'm getting the following error on console:

```
ConnectIntegrationForm.tsx:84 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'toString')
    at onCreatIntegration (ConnectIntegrationForm.tsx:84:1)
    at async createFormControl.ts:1044:1
```

File: providers/sendgrid/src/lib/sendgrid.provider.ts [hunk 4]
```diff
@@ -25,7 +27,40 @@ export class SendgridEmailProvider implements IEmailProvider {
...
+      const mailData = this.createMailData(options);
+
+      const response = await this.sendgridMail.send(mailData);
+
+      if (response[0]?.statusCode === 202) {
+        return {
+          success: true,
+          message: 'Integration Successful',
+        };
+      }
```
Comment (scopsy) [inline providers/sendgrid/src/lib/sendgrid.provider.ts:52]: It looks like in case the message is 200, we will return a void message. Could we just remove this `if` statement and rely on the sendgrid Throw mechanism?

File: providers/sendgrid/src/lib/sendgrid.provider.ts [hunk 5]
```diff
@@ -25,7 +27,40 @@ export class SendgridEmailProvider implements IEmailProvider {
...
+    } catch (error) {
+      return {
+        success: false,
+        message:
```
Comment (scopsy) [inline providers/sendgrid/src/lib/sendgrid.provider.ts:56]: Looks like here we are discarding the provider message at all, could we just return the Enum of our internal error on one field and also return the provider full text in case it exists? I think otherwise we might mask important information from the user. Wdyt? @BiswaViraj

File: apps/web/src/pages/integrations/components/ConnectIntegrationForm.tsx [hunk 6]
```diff
@@ -131,7 +143,25 @@ export function ConnectIntegrationForm({
...
+                render={({ field }) => (
+                  <Switch
+                    checked={isCheckIntegration}
+                    data-test-id="is_check_integration_id"
+                    {...field}
+                    onChange={handlerCheckIntegrationChange}
+                  />
+                )}
+              />
+              <StyledText>Send a test email to verify the credentials?</StyledText>
```
Comment (scopsy) [inline apps/web/src/pages/integrations/components/ConnectIntegrationForm.tsx:160]:
```suggestion
              <StyledText>Verify provider credentials</StyledText>
```

File: packages/stateless/src/lib/provider/provider.interface.ts [hunk 7]
```diff
@@ -59,6 +59,8 @@ export interface IEmailProvider extends IProvider {
   channelType: ChannelTypeEnum.EMAIL;
 
   sendMessage(options: IEmailOptions): Promise<ISendMessageSuccessResponse>;
+
+  checkIntegration(options: IEmailOptions): Promise<ICheckIntegrationResponse>;
```
Comment (scopsy) [inline packages/stateless/src/lib/provider/provider.interface.ts:63]: Why are we passing the email options here? I would suggest that each provider will handle this logic internally, since some of them might actually have a test endpoint. Wdyt?

File: packages/stateless/src/lib/provider/provider.enum.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export enum CheckIntegrationResponseEnum {
+  INVALID_EMAIL = 'From Email id is invalid/incorrect',
```
Comment (scopsy) [inline packages/stateless/src/lib/provider/provider.enum.ts:2]: When doing the error messages enum, I would stick just to an enum value. The display text should be handled in the UI or the server should return the message directly.

```suggestion
  INVALID_EMAIL = 'invalid_email',
```

My suggestion is just to have the enum value here and separate the responsibilities over this one. 

Let's discuss it together later :)

File: providers/sendgrid/src/lib/sendgrid.provider.ts (hunk 5 above)
Comment (BiswaViraj) [reply providers/sendgrid/src/lib/sendgrid.provider.ts:56]: Providers' message is returned when none of the messages our Enum matches

`mapResponse(error?.code) || error?.response?.body?.errors[0]?.message,` handles it

Comment (scopsy) [reply providers/sendgrid/src/lib/sendgrid.provider.ts:56]: What I'm suggesting here, is to **always** return the provider full message and in another field our mapped enum error in case we want to handle it differently. Wdyt?

Comment (BiswaViraj) [reply providers/sendgrid/src/lib/sendgrid.provider.ts:56]: I think that would be good.

`js
return {
 success: false,
 message: provider's response message,
 type: Mapped Enum(INVALID_EMAIL, BAD_CREDENTIALS.....)
`
Does this format look good? wdyt?

Comment (scopsy) [reply providers/sendgrid/src/lib/sendgrid.provider.ts:56]: Perfect! Would maybe just adjust the `type` to be `code`

File: apps/web/src/pages/integrations/components/ConnectIntegrationForm.tsx (hunk 3 above)
Comment (BiswaViraj) [reply apps/web/src/pages/integrations/components/ConnectIntegrationForm.tsx]: Done ;)

**Comment by BiswaViraj** [comment]
[Test_Provider.webm](https://user-images.githubusercontent.com/29251776/191069872-39dd78f9-8aff-4f12-ac78-a7ba3e28afe1.webm)

Create/Update Test provider flow on UI

**Review by scopsy — APPROVED** [review]
